File: tools/s07/cpu-profile/go/main.go
// Diagnostic CPU profiling adapter for the pinned source parser and binder.
// Its queue and loaded-input identity match tools/s07/benchmark/main.go.
// Profiling and per-file timers add overhead: none of these timings or CPU
// sample shares are acceptance measurements.
package main

import (
	"context"
	"crypto/sha256"
	"encoding/binary"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"runtime/debug"
	"runtime/pprof"
	"strconv"
	"sync"
	"time"

	"github.com/microsoft/TypeScript/tsc/internal/ast"
	"github.com/microsoft/TypeScript/tsc/internal/binder"
	"github.com/microsoft/TypeScript/tsc/internal/core"
	"github.com/microsoft/TypeScript/tsc/internal/parser"
	"github.com/microsoft/TypeScript/tsc/internal/tspath"
	"github.com/microsoft/TypeScript/tsc/internal/vfs/osvfs"
)

type Input struct {
	Filename   string `json:"filename"`
	Path       string `json:"path"`
	Local      string `json:"local"`
	ScriptKind int32  `json:"script_kind"`
	JSX        bool   `json:"jsx"`
	Force      bool   `json:"force"`
}

type Loaded struct {
	text    string
	options ast.SourceFileParseOptions
	kind    core.ScriptKind
}

// These sums measure elapsed time inside the actual parse/bind callbacks,
// including scheduler and GC pauses. They are not per-thread CPU time, and
// concurrent workers' sums can exceed the overall phase wall time.
type WorkerTimes struct {
	Worker     int   `json:"worker"`
	Files      int   `json:"files"`
	ParseCalls int   `json:"parse_calls"`
	BindCalls  int   `json:"bind_calls"`
	ParseNs    int64 `json:"parse_ns"`
	BindNs     int64 `json:"bind_ns"`
}

type Report struct {
	Version           int           `json:"version"`
	Operation         string        `json:"operation"`
	DiagnosticOnly    bool          `json:"diagnostic_only"`
	LoadedInputSHA256 string        `json:"loaded_input_sha256"`
	Workers           int           `json:"workers"`
	Files             int           `json:"files"`
	LoadedBytes       int           `json:"loaded_bytes"`
	Nodes             int           `json:"nodes"`
	Symbols           int           `json:"symbols"`
	ParseDiagnostics  int           `json:"parse_diagnostics"`
	BindDiagnostics   int           `json:"bind_diagnostics"`
	WallTimeNs        int64         `json:"wall_time_ns"`
	ParseCalls        int           `json:"parse_calls"`
	BindCalls         int           `json:"bind_calls"`
	ParseNs           int64         `json:"parse_ns"`
	BindNs            int64         `json:"bind_ns"`
	WorkerTimes       []WorkerTimes `json:"worker_times"`
	TimingDomain      string        `json:"timing_domain"`
	ProfileScope      string        `json:"profile_scope"`
	CPUProfile        string        `json:"cpu_profile"`
	StartupNs         int64         `json:"startup_ns"`
	PreloadNs         int64         `json:"preload_ns"`
	WorkerSetupNs     int64         `json:"worker_setup_ns"`
	CPUCapacity       int           `json:"cpu_capacity"`
	GoroutinesReady   int           `json:"goroutines_ready"`
	GOMAXPROCS        int           `json:"gomaxprocs"`
	GOGC              int           `json:"gogc"`
}

// Observe exactly the same immutable preload and byte protocol as the baseline.
// This hash is calculated before profiling or releasing any worker.
func loadedDigest(inputs []Loaded) string {
	digest := sha256.New()
	digest.Write([]byte("S07-loaded-inputs-v1\x00"))
	word := make([]byte, 8)
	binary.BigEndian.PutUint64(word, uint64(len(inputs)))
	digest.Write(word)
	for _, input := range inputs {
		for _, value := range []string{input.options.FileName, string(input.options.Path)} {
			binary.BigEndian.PutUint64(word, uint64(len(value)))
			digest.Write(word)
			digest.Write([]byte(value))
		}
		binary.BigEndian.PutUint32(word[:4], uint32(input.kind))
		digest.Write(word[:4])
		bits := []byte{0, 0}
		if input.options.ExternalModuleIndicatorOptions.JSX {
			bits[0] = 1
		}
		if input.options.ExternalModuleIndicatorOptions.Force {
			bits[1] = 1
		}
		digest.Write(bits)
		binary.BigEndian.PutUint64(word, uint64(len(input.text)))
		digest.Write(word)
		source := sha256.Sum256([]byte(input.text))
		digest.Write(source[:])
	}
	return fmt.Sprintf("%x", digest.Sum(nil))
}

func main() {
	entered := time.Now()
	if len(os.Args) != 4 {
		panic("usage: cpu-profile INPUTS.json WORKERS (1 or 8) OUTPUT_PREFIX")
	}
	workers, err := strconv.Atoi(os.Args[2])
	if err != nil || workers != 1 && workers != 8 {
		panic("worker count must be 1 or 8")
	}
	if os.Args[3] == "" {
		panic("empty output prefix")
	}
	input, err := os.Open(os.Args[1])
	if err != nil {
		panic(err)
	}
	decoder := json.NewDecoder(input)
	decoder.DisallowUnknownFields()
	var requests []Input
	if err := decoder.Decode(&requests); err != nil {
		panic(err)
	}
	var trailing any
	if err := decoder.Decode(&trailing); err != io.EOF {
		panic("workload must contain exactly one JSON value")
	}
	if err := input.Close(); err != nil {
		panic(err)
	}
	if len(requests) == 0 {
		panic("empty parse-and-bind workload")
	}
	runtime.GOMAXPROCS(workers)
	debug.SetGCPercent(100)
	startupNs := time.Since(entered).Nanoseconds()
	preloadStarted := time.Now()
	inputs := make([]Loaded, len(requests))
	loadedBytes := 0
	for index, request := range requests {
		text, ok := osvfs.FS().ReadFile(request.Local)
		if !ok {
			panic("could not load workload file: " + request.Local)
		}
		inputs[index] = Loaded{text, ast.SourceFileParseOptions{FileName: request.Filename, Path: tspath.Path(request.Path), ExternalModuleIndicatorOptions: ast.ExternalModuleIndicatorOptions{JSX: request.JSX, Force: request.Force}}, core.ScriptKind(request.ScriptKind)}
		loadedBytes += len(text)
	}
	loadedInputSHA256 := loadedDigest(inputs)
	preloadNs := time.Since(preloadStarted).Nanoseconds()

	// Open both outputs before the phase. A phase failure leaves no successful
	// report at this prefix; its partial profile can still help diagnose failure.
	prefix, err := filepath.Abs(os.Args[3])
	if err != nil {
		panic(err)
	}
	if err := os.MkdirAll(filepath.Dir(prefix), 0o755); err != nil {
		panic(err)
	}
	reportFile, err := os.Create(prefix + ".report.json")
	if err != nil {
		panic(err)
	}
	defer reportFile.Close()
	profileFile, err := os.Create(prefix + ".pprof")
	if err != nil {
		panic(err)
	}
	defer profileFile.Close()

	workerSetupStarted := time.Now()
	channels := make([]chan int, workers)
	results := make([][]*ast.SourceFile, workers)
	timings := make([]WorkerTimes, workers)
	failures := make([]any, workers)
	ready := sync.WaitGroup{}
	ready.Add(workers)
	done := sync.WaitGroup{}
	done.Add(workers)
	start := make(chan struct{})
	for worker := range workers {
		channels[worker] = make(chan int, workers)
		results[worker] = make([]*ast.SourceFile, 0, (len(inputs)+workers-1)/workers)
		timings[worker].Worker = worker
		go func(worker int) {
			defer done.Done()
			ctx := context.Background()
			parseLabels := pprof.Labels("phase", "parse")
			bindLabels := pprof.Labels("phase", "bind")
			ready.Done()
			<-start
			for index := range channels[worker] {
				if failures[worker] != nil {
					// Drain after a panic so the bounded dispatcher can finish.
					continue
				}
				func() {
					completed := false
					defer func() {
						failure := recover()
						if !completed {
							failures[worker] = fmt.Sprintf("file %s: %v\n%s", inputs[index].options.FileName, failure, debug.Stack())
						}
					}()
					input := inputs[index]
					var file *ast.SourceFile
					pprof.Do(ctx, parseLabels, func(context.Context) {
						started := time.Now()
						file = parser.ParseSourceFile(input.options, input.text, input.kind)
						timings[worker].ParseNs += time.Since(started).Nanoseconds()
						timings[worker].ParseCalls++
					})
					pprof.Do(ctx, bindLabels, func(context.Context) {
						started := time.Now()
						binder.BindSourceFile(file)
						timings[worker].BindNs += time.Since(started).Nanoseconds()
						timings[worker].BindCalls++
					})
					results[worker] = append(results[worker], file)
					timings[worker].Files++
					completed = true
				}()
			}
		}(worker)
	}
	ready.Wait()
	workerSetupNs := time.Since(workerSetupStarted).Nanoseconds()
	cpuCapacity, goroutinesReady := runtime.NumCPU(), runtime.NumGoroutine()
	if err := pprof.StartCPUProfile(profileFile); err != nil {
		panic(err)
	}
	profiling := true
	defer func() {
		if profiling {
			pprof.StopCPUProfile()
		}
	}()
	started := time.Now()
	close(start)
	for index := range inputs {
		channels[index%workers] <- index
	}
	for _, channel := range channels {
		close(channel)
	}
	done.Wait()
	elapsed := time.Since(started)
	pprof.StopCPUProfile()
	profiling = false
	if err := profileFile.Close(); err != nil {
		panic(err)
	}

	// Roots remain reachable through profile shutdown and the entire summary.
	// No GC, semantic traversal, serialization, or root release is added between
	// the worker endpoint and StopCPUProfile.
	report := Report{
		Version: 1, Operation: "parse_bind_cpu_profile", DiagnosticOnly: true,
		LoadedInputSHA256: loadedInputSHA256, Workers: workers, LoadedBytes: loadedBytes,
		WallTimeNs: elapsed.Nanoseconds(), WorkerTimes: timings,
		TimingDomain: "parse_ns and bind_ns sum per-file elapsed worker time inside the labeled callbacks, including scheduler and GC pauses; they exclude pprof label setup/restore and are not CPU time",
		ProfileScope: "CPU samples from before the worker barrier release until immediately after all workers finish; pprof.Do phase=parse surrounds ParseSourceFile and phase=bind surrounds BindSourceFile; labels include their wrapper overhead, while runtime/background work may be unlabeled",
		CPUProfile:   profileFile.Name(), GOMAXPROCS: runtime.GOMAXPROCS(0), GOGC: 100,
		StartupNs: startupNs, PreloadNs: preloadNs, WorkerSetupNs: workerSetupNs,
		CPUCapacity: cpuCapacity, GoroutinesReady: goroutinesReady,
	}
	for worker, files := range results {
		if failures[worker] != nil {
			panic(failures[worker])
		}
		expected := len(inputs) / workers
		if worker < len(inputs)%workers {
			expected++
		}
		stats := timings[worker]
		if len(files) != expected || stats.Files != expected || stats.ParseCalls != expected || stats.BindCalls != expected {
			panic(fmt.Sprintf("worker %d workload count mismatch: expected %d, retained %d, completed %+v", worker, expected, len(files), stats))
		}
		report.ParseCalls += stats.ParseCalls
		report.BindCalls += stats.BindCalls
		report.ParseNs += stats.ParseNs
		report.BindNs += stats.BindNs
		for _, file := range files {
			report.Files++
			report.Nodes += file.NodeCount
			report.Symbols += file.SymbolCount
			report.ParseDiagnostics += len(file.Diagnostics())
			report.BindDiagnostics += len(file.BindDiagnostics())
		}
	}
	if report.Files != len(inputs) || report.ParseCalls != len(inputs) || report.BindCalls != len(inputs) {
		panic("workload file count mismatch")
	}
	if err := json.NewEncoder(reportFile).Encode(report); err != nil {
		panic(err)
	}
	if err := reportFile.Close(); err != nil {
		panic(err)
	}
	if err := json.NewEncoder(os.Stdout).Encode(report); err != nil {
		panic(err)
	}
	runtime.KeepAlive(results)
}

File: tools/s07/cpu-profile/rust/src/main.rs
//! Phase attribution for the unchanged S07 parse/bind workload. This adapter
//! emits diagnostic elapsed timers; Time Profiler supplies separate CPU samples.
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};
use std::{
    fs,
    hint::black_box,
    io::{self, Write},
    path::Path,
    sync::{mpsc, Barrier},
    thread,
    time::Instant,
};
use ts_ast::{
    AstFile, BoundFile, ExternalModuleIndicatorOptions, JsString, NodeId, ParsedFile,
    SourceFileParseOptions,
};
use ts_jsstring::SourceText;

#[global_allocator]
static ALLOCATOR: mimalloc::MiMalloc = mimalloc::MiMalloc;

#[derive(Deserialize)]
#[serde(deny_unknown_fields)]
struct Input {
    filename: String,
    path: String,
    local: String,
    script_kind: i32,
    jsx: bool,
    force: bool,
}
struct Loaded {
    source: SourceText,
    options: SourceFileParseOptions,
    script_kind: ts_core::ScriptKind,
}
#[derive(Default, Serialize)]
struct WorkerTimes {
    parse_ns: u128,
    publish_ns: u128,
    bind_ns: u128,
}
#[derive(Serialize)]
struct Report {
    version: u32,
    runtime: &'static str,
    diagnostic_only: bool,
    workers: usize,
    files: usize,
    loaded_bytes: usize,
    loaded_input_sha256: String,
    nodes: i64,
    symbols: isize,
    parse_diagnostics: usize,
    bind_diagnostics: usize,
    pipeline_wall_ns: u128,
    worker_times: Vec<WorkerTimes>,
    elapsed_worker_totals: WorkerTimes,
    timer_domain: &'static str,
}

#[inline(never)]
fn profile_preload(path: &Path) -> Result<Vec<Loaded>, Box<dyn std::error::Error>> {
    let inputs: Vec<Input> = serde_json::from_slice(&fs::read(path)?)?;
    if inputs.is_empty() {
        return Err("empty workload".into());
    }
    inputs
        .into_iter()
        .map(|input| {
            Ok(Loaded {
                source: SourceText::from_bytes(fs::read(input.local)?),
                options: SourceFileParseOptions {
                    file_name: JsString::from_bytes(input.filename.into_bytes()),
                    path: JsString::from_bytes(input.path.into_bytes()),
                    external_module_indicator_options: ExternalModuleIndicatorOptions {
                        jsx: input.jsx,
                        force: input.force,
                    },
                },
                script_kind: ts_core::ScriptKind(input.script_kind),
            })
        })
        .collect()
}

fn loaded_digest(inputs: &[Loaded]) -> String {
    let mut digest = Sha256::new();
    digest.update(b"S07-loaded-inputs-v1\0");
    digest.update((inputs.len() as u64).to_be_bytes());
    for input in inputs {
        for value in [
            input.options.file_name.as_bytes(),
            input.options.path.as_bytes(),
        ] {
            digest.update((value.len() as u64).to_be_bytes());
            digest.update(value);
        }
        digest.update(input.script_kind.0.to_be_bytes());
        digest.update([
            u8::from(input.options.external_module_indicator_options.jsx),
            u8::from(input.options.external_module_indicator_options.force),
        ]);
        digest.update((input.source.len() as u64).to_be_bytes());
        digest.update(Sha256::digest(input.source.as_bytes()));
    }
    format!("{:x}", digest.finalize())
}

// Named, non-inlined frames preserve the original interleaved operation order.
// black_box prevents a tail call from erasing the phase frame in sampled stacks.
#[inline(never)]
fn profile_parse(input: &Loaded) -> ParsedFile {
    black_box(ts_parser::parse_source_file(
        input.source.clone(),
        input.script_kind,
        input.options.clone(),
    ))
}

#[inline(never)]
fn profile_publish(parsed: ParsedFile) -> (AstFile, NodeId) {
    let source = parsed.root();
    black_box((parsed.publish_unbound(), source))
}

#[inline(never)]
fn profile_bind(file: AstFile, source: NodeId) -> BoundFile {
    let bound = ts_binder::bind_source_file(&file, source).expect("workload binding must complete");
    drop(file);
    black_box(bound)
}

#[inline(never)]
fn measure_pipeline(
    senders: Vec<mpsc::SyncSender<usize>>,
    count: usize,
    finished: &Barrier,
) -> u128 {
    let start = Instant::now();
    for index in 0..count {
        senders[index % senders.len()]
            .send(index)
            .expect("worker drains every input");
    }
    drop(senders);
    finished.wait();
    black_box(start.elapsed().as_nanos())
}

#[inline(never)]
fn profile_retirement(files: Vec<Vec<BoundFile>>) {
    drop(black_box(files));
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<_> = std::env::args_os().collect();
    if args.len() != 3 {
        return Err("usage: ts_cpu_profile INPUTS.json WORKERS (1 or 8)".into());
    }
    let workers: usize = args[2].to_str().ok_or("invalid workers")?.parse()?;
    if !matches!(workers, 1 | 8) || thread::available_parallelism()?.get() < workers {
        return Err("profiling requires 1 or 8 available workers".into());
    }
    let inputs = profile_preload(Path::new(&args[1]))?;
    let file_count = inputs.len();
    let loaded_bytes = inputs.iter().map(|input| input.source.len()).sum();
    let loaded_input_sha256 = loaded_digest(&inputs);
    let ready = Barrier::new(workers + 1);
    let finished = Barrier::new(workers + 1);
    let release = Barrier::new(workers + 1);
    thread::scope(|scope| -> Result<(), Box<dyn std::error::Error>> {
        let mut senders = Vec::with_capacity(workers);
        let mut handles = Vec::with_capacity(workers);
        for _ in 0..workers {
            let (send, receive) = mpsc::sync_channel::<usize>(workers);
            senders.push(send);
            let inputs = &inputs;
            let (ready, finished, release) = (&ready, &finished, &release);
            handles.push(ts_parser::spawn_parser_worker(scope, move || {
                let mut roots = Vec::with_capacity(file_count.div_ceil(workers));
                let mut times = WorkerTimes::default();
                let mut failure = None;
                ready.wait();
                while let Ok(index) = receive.recv() {
                    if failure.is_some() {
                        continue;
                    }
                    let outcome = std::panic::catch_unwind(std::panic::AssertUnwindSafe(|| {
                        let start = Instant::now();
                        let parsed = profile_parse(&inputs[index]);
                        times.parse_ns += start.elapsed().as_nanos();
                        let start = Instant::now();
                        let (file, source) = profile_publish(parsed);
                        times.publish_ns += start.elapsed().as_nanos();
                        let start = Instant::now();
                        let bound = profile_bind(file, source);
                        times.bind_ns += start.elapsed().as_nanos();
                        bound
                    }));
                    match outcome {
                        Ok(file) => roots.push(file),
                        Err(panic) => failure = Some(panic),
                    }
                }
                finished.wait();
                release.wait();
                if let Some(panic) = failure {
                    std::panic::resume_unwind(panic);
                }
                (roots, times)
            })?);
        }
        ready.wait();
        let pipeline_wall_ns = measure_pipeline(senders, file_count, &finished);
        release.wait();
        let mut files = Vec::with_capacity(workers);
        let mut worker_times = Vec::with_capacity(workers);
        let mut totals = WorkerTimes::default();
        for handle in handles {
            let (roots, times) = handle
                .join()
                .unwrap_or_else(|panic| std::panic::resume_unwind(panic));
            totals.parse_ns += times.parse_ns;
            totals.publish_ns += times.publish_ns;
            totals.bind_ns += times.bind_ns;
            worker_times.push(times);
            files.push(roots);
        }
        let mut report = Report {
            version: 1,
            runtime: "rust",
            diagnostic_only: true,
            workers,
            files: 0,
            loaded_bytes,
            loaded_input_sha256,
            nodes: 0,
            symbols: 0,
            parse_diagnostics: 0,
            bind_diagnostics: 0,
            pipeline_wall_ns,
            worker_times,
            elapsed_worker_totals: totals,
            timer_domain:
                "sum of per-file elapsed worker timers; includes descheduling; not CPU time",
        };
        for file in files.iter().flatten() {
            let view = file.view();
            let source = view.source_file()?;
            report.files += 1;
            report.nodes += source.node_count;
            report.symbols += view.result().symbol_count();
            report.parse_diagnostics += source.diagnostics.len();
            report.bind_diagnostics += view.result().diagnostics().len();
        }
        if report.files != file_count {
            return Err("incomplete workload".into());
        }
        println!("{}", serde_json::to_string(&report)?);
        io::stdout().flush()?;
        black_box(&files);
        profile_retirement(files);
        Ok(())
    })
}
